Block Editor: Try using DependencyExtractionWebpackPlugin

gziolo opened the pull request:
Work in progress.

I'll move to Trac as soon as Travis turns green :)

sirreal commented:
You'll want to delete these lines:

https://github.com/WordPress/wordpress-develop/blob/f8688bb43b8e9a9478de356c086dfdab11bd1f77/tools/webpack/packages.js#L110-L124

gziolo commented:
@sirreal, thanks for pointers. It solved the issue 🎉 

I had to include `@wordpress/block-directory` as a dependency. We will have to fix it on Gutenberg side first as it's hidden behind the flag and it shouldn't be necessary to be registered and enqueued in WordPress 5.3. /cc @youknowriad 

@omarreiss, we add a number of PHP files for each JS script. Is it fine to keep them next to distribution version of JS files as presented on the screenshot?

![Screen Shot 2019-09-24 at 17 27 29](https://user-images.githubusercontent.com/699132/65526042-b7e0ef80-def0-11e9-8c5d-9e1ab19053a1.png)

The same structure is present in both `build` and `src` folders.

gziolo commented:
The discussion should continue on WordPress Trac: https://core.trac.wordpress.org/ticket/48154.

gziolo commented:
I updated the patch with the latest changes from the `master` branch. 

In addition, I switched the output format for asset files to JSON.

![Screen Shot 2019-12-04 at 11 23 24](https://user-images.githubusercontent.com/699132/70134500-81f67f00-1688-11ea-83be-93c6e31e5efe.png)

It also looks like JSON is now required to run WordPress so we can take advantage of it:
https://make.wordpress.org/core/2019/10/15/php-native-json-extension-now-required/

gziolo commented on Gruntfile.js:
```diff
@@ -342,6 +348,12 @@ module.exports = function(grunt) {
+			assets: {
+				expand: true,
+				cwd: WORKING_DIR + 'wp-includes/js/',
+    			src: 'dist/*.asset.php',
```
hmmm, I need to fix it...

ntwb commented on Gruntfile.js:
```diff
@@ -108,8 +108,12 @@ module.exports = function(grunt) {
 				WORKING_DIR + 'wp-includes/js/'
 			],
+			assets: [
+				WORKING_DIR + 'wp-includes/js/**/*.asset.php'
```
Based [on the core trac image](https://core.trac.wordpress.org/attachment/ticket/48154/Screen%20Shot%202019-12-04%20at%2016.10.08.png) the `php` assets are now stored in the `/src/wp-includes/assets/dist` folder

gziolo replied:
It's more complex than that.

This is where webpack generates those PHP assets, we move them to `/src/wp-includes/assets/dist` or `/build/wp-includes/assets/dist` depending where webpack operates. It's even more hard to follow when you take into account that `npm run build:dev`  operates in `src` folder, but `npm run build` uses `build` folder ...

gziolo replied:
This name should probably be updated to `webpack-generated-assets` :)

gziolo replied:
@ntwb I included some tweaks to make it easier to follow. I hope it helps:

https://github.com/WordPress/wordpress-develop/pull/102/commits/ba91cc12103ffc7d0720c0c9be0aa850d47ce5cd

gziolo commented:
Resolved with https://core.trac.wordpress.org/changeset/47035.
